### ISSUE TITLE
Critical Authorization Failure with Claim handling

### ISSUE THREAD

**Issue by optimizasean** [dotnet/runtime#55732]
Using .NET6 preview 5 with Blazor Web Assembly:

When dealing with claims, if there are multiple claims with the same types, their values get concatenated into an "array-like object".  I specify this because they are not some form of IEnumerable<string> or string[] or anything similar.  Here is the result:

3 claims with same type:
Claim: { type="admin", value="a" }
Claim: { type="admin", value="45" }
Claim: { type="admin", value="184927-6a" }

Doesn't get combined into this: Claim:{ string::type="admin", (somekindof)string[]::value=["a","45", "184927-6a"] }

It actually gets combined into this: Claim:{ string::type="admin", string::value="[\"a\",\"45\", \"184927-6a\"]" }

This makes it pretty difficult to work with when the default methods fail to detect the claim values for matching claims.
When you call HasClaim on the above, it fails always if you check the value.
With value of "a", value of ["a","45", "184927-6a"], or even value of "[\"a\",\"45\", \"184927-6a\"]" with different encodings it always fails.

The intention of HasClaim<string, string> should be to find a Claim that has a specific type and a matching value.  This does not seem to work both on ClaimsPrincipal or on ClaimsIdentity.  Use case is for claims based authorization, especially when writing authorization policies in a resource based authorization model such as:
```c#
public static class AuthorizationPolicyBuilderExtensions {
        public static AuthorizationPolicyBuilder RequireResourceClaim(this AuthorizationPolicyBuilder authPolicyBuilder, string claimType, int resourceLocation) => authPolicyBuilder.RequireAssertion(
            (AuthorizationHandlerContext auth) => {
                if (auth.Resource is NavigationManager navMan) {
                    string[] route = navMan.ToBaseRelativePath(navMan.Uri).Split('/');
                    int loc = route.Length - resourceLocation;
                    if (loc < 0 || loc >= route.Length) return false;
                    string resourceValue = route?[loc];
                    return auth.User.HasClaim(claimType, resourceValue);
                } else return false;
            }
        );
    }
```
The above I have attached to assist anyone attempting to create a resource based authorization model (which would be nice supported with URL parameters through attribute like `@attribute [Authorize(policy, resource, ....)]`).

It can be used in the following scenario:

Razor page to use it:
```razor
@page "/testing/resauth/{data}/action"
@page "/testing/resauth/action/{data}"

@inject NavigationManager navMan

<h1 class="text-4xl">Auth Res</h1>

<AuthorizeView Policy="Testing" Resource="navMan">
    <Authorized>
        <p>Authorized</p>
    </Authorized>
    <NotAuthorized>
        <p>Failure</p>
    </NotAuthorized>
</AuthorizeView>
```

Policies.cs:
```c#
public static class Policies {
    public const string Testing = "Testing";
    public static AuthorizationPolicy TestingPolicy() => new AuthorizationPolicyBuilder().RequireAuthenticatedUser().RequireResourceClaim("Team Admin User", 2).Build();
}
```

Program.cs:
```c#
public class Program {
    public static async Task Main(string[] args) {
        var builder = WebAssemblyHostBuilder.CreateDefault(args);
        builder.Services.AddOptions();
        builder.Services.AddAuthorizationCore(options => {
            options.AddPolicy(Policies.Testing, Policies.TestingPolicy());
        });
    }
}
```

In summary there should be documentation covering the default behavior of claims (joining values), the HasClaim method needs to be fixed as since the joining of claims is a framework behavior then HasClaim should be able to deal with that, and there should be something in place to check for this if some test did not catch this error.

Seriously frustrating issue here, I really appreciate the help.  Able to run it down this far thanks to some epic folks at the AspNet gitter!

**Comment by radical** [comment]
@optimizasean Is this a blazor/wasm specific issue? Or does this happen in a non-blazor/wasm app too? It would help with getting the right people looking at this.

I tried the code from https://github.com/dotnet/runtime/issues/55732#issuecomment-880381221 (try.dot.net) in a (non-blazor) wasm app, and got the same output.

**Comment by radical** [comment]
@captainsafia Could this be a marshaling issue?

**Comment by captainsafia** [comment]
I don't suspect so. It seems like the feedback here is about the implementation of the HasClaim API.

https://github.com/dotnet/runtime/blob/016851a183c66333f3087312b458452a90154726/src/libraries/System.Security.Claims/src/System/Security/Claims/ClaimsPrincipal.cs#L426

I suspect the issue was moved here because the suspected bug is in the runtime library and not related to WASM.

**Comment by bartonjs** [comment]
@optimizasean In your code snippet (in https://github.com/dotnet/runtime/issues/55732#issuecomment-880381221) you don't ever use the `claims` list.  Multi-claim and claims-actual are the same because they're actually both sourcing the built multi-claim claim.

Changing `var claimsIdentity = new ClaimsIdentity(new List<Claim>() { multiClaim })` to `var claimsIdentity = new ClaimsIdentity(claims)` makes the rest of the sample fail, because it faithfully shows that there were mutiple claims with that same type.

I don't think this is a ClaimsIdentity problem, but rather whatever specific identity type is being used... I'm guessing it's a JWT type, because the merging behavior is something that JwtPayload does:  https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/824068adaeb7b158df2b36d8787f689646e715e0/src/System.IdentityModel.Tokens.Jwt/JwtPayload.cs#L513-L548, and JWT is popular these days.

I don't think there's anything for dotnet/runtime to do here.  It's really down to who picked the ClaimsIdentity-derived type that has this behavior (ASP.NET?) or the type itself.

**Comment by optimizasean** [comment]
Oops @barton, nice catch!  fixing source...

**Comment by optimizasean** [comment]
@bartonjs You are right, it is not the ClaimsIdentity or ClaimsPrincipal that does the bundling.  It is the system JWT package.  I use JWT in my application.  I bundle the claims in my example to represent how it ends up after being processed by .NET (System.IdentityModel.Tokens.Jwt) which is to combine all claims of same type into value array.  It seems to be the Microsoft intended behavior according to: https://github.com/dotnet/runtime/pull/46464, where this is implemented in HasClaim (fixed at some point possibly?).  There are also referenced issues with identity server doing the same here: https://github.com/aspnet/Identity/issues/1394

The main problem is I am using all default .NET nuget packages (Microsoft origin only) or built in System packages and it doesn't play well with itself since it doesn't exactly deserialize nicely as a JSON array either using System.Text.Json.  Also with the prevalence of JSON/SAML/... and other authentication types, especially since essentially every Blazor WASM application uses JSON at some point, .NET should have support for JWT tokens or at a minimum the documentation on the accepted way to handle this since this is how the built in JWT is built by default using .NET like you said with the JwtPayload.

**Comment by optimizasean** [comment]
> @optimizasean Is this a blazor/wasm specific issue? Or does this happen in a non-blazor/wasm app too? It would help with getting the right people looking at this.
> 
> I tried the code from [#55732 (comment)](https://github.com/dotnet/runtime/issues/55732#issuecomment-880381221) (try.dot.net) in a (non-blazor) wasm app, and got the same output.

It is related to the JWT built through System.IdentityModel.Tokens.Jwt.  JwtPayload apparently combines all claims of same type into a string array of values under one claim.  I think this actually should be related to Blazor WASM and how HasClaim is implemented there but this also might be a problem with webapi on the receiving end of JWT tokens.  If HasClaim can't handle that in blazorwasm, I would also assume the same for webapi.

Solutions:
1. HasClaim update to make it able to handle it?
2. Should the system JWT package not combine claims?
3. Documentation on accepted solution / how to fix how system JWT combines claims (feature flag to disable or options maybe?)
4. System.Text.Json make it parse nicely by default (doesn't like format?)
5. Something else entirely?

Depending on which direction this should be taken, should the issue be moved again?  I think it belongs in runtime if the solution is to fix the old HasClaim PR in https://github.com/dotnet/runtime/pull/46464 which would be ideal for .NET users.

**Comment by bartonjs** [comment]
The problem that I had with the HasClaim PR is still present: It's codifying a behavior from one of the derived types, but it's not a behavior that's central to ClaimsPrincipal.

Without understanding why JwtPayload chose to merge them in the first place, it does feel like "don't merge them" is the right answer... but one man's fix is another man's break (see also your snippet that used SingleOrDefault).  That'd probably belong in AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet.

Docs seem reasonable, but I have no idea where it'd be documented... but I'll stop talking technical and tag @blowdart so he can talk scenario.

**Comment by blowdart** [comment]
Yea, maybe asking the identitymodel folks why they're combining is a good start. If it's horribly blocking there are events that you could hook into to unpack it into individual claims again during the aspnet auth process, but again, it's in the identitymodel code rather than framework code.

**Comment by bartonjs** [comment]
Since this is specific to the JWT-based claims provider, you'll have to follow up with the group that produces it (https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet).
